### ISSUE TITLE
GuzzleHandler only handles Exceptions and not Errors

### ISSUE THREAD

**Issue by afrowuk** [aws/aws-sdk-php#2144]
**Describe the bug**
We've seen this error a few times when our servers have a temporary loss of connectivity. It seems that this is resulting in a runtime error that the GuzzleHandler class can't handle, which is masking the underlying runtime error.

This is the error:
`Argument 1 passed to Aws\Handler\GuzzleV6\GuzzleHandler::Aws\Handler\GuzzleV6\{closure}() must be an instance of Exception, instance of Error given, called in /var/www/pathto/vendor/guzzlehttp/promises/src/Promise.php on line 204`

Stack trace:
```
#0 /var/www/pathto/vendor/guzzlehttp/promises/src/Promise.php(204): Aws\Handler\GuzzleV6\GuzzleHandler::Aws\Handler\GuzzleV6\{closure}(Object(Error))
#1 /var/www/pathto/vendor/guzzlehttp/promises/src/Promise.php(153): GuzzleHttp\Promise\Promise::callHandler(2, Object(Error), NULL)
#2 /var/www/pathto/vendor/guzzlehttp/promises/src/TaskQueue.php(48): GuzzleHttp\Promise\Promise::GuzzleHttp\Promise\{closure}()
#3 /var/www/pathto/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(118): GuzzleHttp\Promise\TaskQueue->run()
#4 /var/www/pathto/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(145): GuzzleHttp\Handler\CurlMultiHandler->tick()
#5 /var/www/pathto/vendor/guzzlehttp/promises/src/Promise.php(248): GuzzleHttp\Handler\CurlMultiHandler->execute(true)
#6 /var/www/pathto/vendor/guzzlehttp/promises/src/Promise.php(224): GuzzleHttp\Promise\Promise->invokeWaitFn()
#7 /var/www/pathto/vendor/guzzlehttp/promises/src/Promise.php(269): GuzzleHttp\Promise\Promise->waitIfPending()
(stack trace shows a few loops with waitIfPending() calls)
```

**Version of AWS SDK for PHP?**
3.155.2

**Version of PHP (`php -v`)?**
7.2.33 running on Ubuntu 16.04.1

**To Reproduce (observed behaviour)**
Any AWS API which uses Guzzle that results in a runtime error.

**Expected behaviour**
We should see the actual underlying runtime error.

**Comment by ajredniwja** [comment]
Hey @afrowuk thanks for opening this issue, I will have someone from the team work on it according to the priority decided by the team, you are welcomed to open a PR for team to review if you'd like to work on it.

**Comment by Muchenja** [comment]
We've also been getting the same error with our Moodle site - Exception - Argument 1 passed to Aws\Handler\GuzzleV6\GuzzleHandler::Aws\Handler\GuzzleV6\{closure}() must be an instance of Exception, instance of TypeError given, called in [dirroot]/local/aws/sdk/GuzzleHttp/Promise/Promise.php on line 203More information about this error

**Comment by shiftedreality** [comment]
Gettings the same for https://github.com/magento/magento2

**Comment by SamRemis** [comment]
@afrowuk Sorry for the delay, we have a PR in now that should fix the issue :) the error will still be packed inside a RejectedException because it originated in a promise, but it gracefully catches the error

**Comment by SamRemis** [comment]
This was fixed by PR #2163 - if you're experiencing this issue in a version greater than 3.171.0, please reopen this issue or make a new one
